lang/src/org/partiql/lang/planner/transforms/AstToLogicalVisitorTransform.kt

```diff
@@ -81,6 +87,73 @@ internal class AstToLogicalVisitorTransform(
         return expr
     }
 
+    /**
+     * Transforms the input [node] into a pair of two items:
+     *  1. An AST (where all [PartiqlAst.Expr.CallWindow]'s in the input projection list are replaced with references
+     *   ([PartiqlAst.Expr.Id]) to new [PartiqlLogical.VarDecl]'s)
+     *  2. A modified algebra such that each window function call in the input projection list corresponding to one window operator
+     *  TODO: if multiple window functions are operating on the same window, we can potentially add them in a single window operator to increase performance
+     *
+     * Transforms a query like:
+     * ```
+     * SELECT LAG(t.a) OVER (...), LEAD(t.a) OVER (...)
+     * FROM t AS t
+     * ```
+     * into a logical plan similar to:
+     * ```
+     * bindings_to_values
+     *   struct ...
+     *   window
+     *      window
+     *          scan
+     *          over
+     *          window_expression
+     *              $__partiql_window_function_0
+     *              LAG
+     *              t.a
+     *      over
+     *      window_expression
+     *          $__partiql_window_function_1
+     *          LEAD
+     *          t.a
+     * ```
+     */
+    private fun transformWindowFunctions(node: PartiqlAst.Expr.Select, algebra: PartiqlLogical.Bexpr): Pair<PartiqlAst.Expr.Select, PartiqlLogical.Bexpr>? {
+        val windowReplacer = CallWindowFunctionProjectionReplacer()
+
+        val transformedNode = windowReplacer.transformExprSelect(node) as PartiqlAst.Expr.Select
+
+        val windowExpressions = windowReplacer.getWindowFuncs()
```

Comment:
This is essentially where the scoping issue mentioned before gets messed up. We've accidentally recursed too far, so window functions are created at the top-level, then we recurse and the LAG/LEAD is gone and the ID has taken its place.